core/src/storage/cache.py

```diff
@@ -1,36 +1,83 @@
 from trezor.crypto import random
 
 if False:
-    from typing import Optional
+    from typing import Optional, Dict, List
 
+_MAX_SESSIONS_COUNT = 10
+_SESSIONLESS_FLAG = 128
+
+# Traditional cache keys
 APP_COMMON_SEED = 0
-APP_COMMON_SEED_WITHOUT_PASSPHRASE = 1
-APP_CARDANO_ROOT = 2
-APP_MONERO_LIVE_REFRESH = 3
+APP_CARDANO_ROOT = 1
+APP_MONERO_LIVE_REFRESH = 2
+
+# Keys that are valid across sessions
+APP_COMMON_SEED_WITHOUT_PASSPHRASE = 1 | _SESSIONLESS_FLAG
+
 
-_cache_session_id = None  # type: Optional[bytes]
-_cache = {}
+_active_session_id = None  # type: Optional[bytes]
+_caches = {}  # type: Dict[bytes, Dict[int, Any]]
+_session_ids = []  # type: List[bytes]
+_sessionless_cache = {}  # type: Dict[int, Any]
 
 if False:
     from typing import Any
 
 
-def get_session_id() -> bytes:
-    global _cache_session_id
-    if not _cache_session_id:
-        _cache_session_id = random.bytes(32)
-    return _cache_session_id
+def _move_session_ids_queue(session_id: bytes) -> None:
+    # Move the LRU session ids queue.
+    if session_id in _session_ids:
+        _session_ids.remove(session_id)
+    else:
```

Comment:
this `else` should not be here IMO

Reply:
You mean because the `while` loop is not evaluated in that case? It still seems more readable if present imho

Reply:
the `while` loop is defending for a case where `len(_session_ids) > _MAX_SESSIONS_COUNT`
that can be true even in case we removed an existing session id

Reply:
Hm I guess if we have the `while` there we indeed should do it as you suggest. Will fix.